### PR TITLE
Support for Rockchip's family SoCs

### PR DESCRIPTION
**Pull request by rkchrome**
This patch adds to support the RK3368 and RK3399 SoCs.

RK3368/RK3399 is one of the Rockchip family SoCs, which is an
multi-cores ARM SoCs.

This patch adds support to boot the Trusted Firmware on RK3368/RK3399
SoCs, and adds support to boot secondary CPUs, enter/exit core
power states for all CPUs in the slow/fast clusters.

This is the initial version for rockchip SoCs.(RK3368/RK3399 and next SoCs)
* Support arm gicv2 & gicv3.
* Boot up multi-cores CPU.
* Add generic CPU helper functions.
* Support suspend/resume.
* Add system_off & system_reset implementation.
* Add delay timer platform implementation.
* Support the new porting interface for the PSCI implementation.

Change-Id: I704bb3532d65e8c70dbd99b512c5e6e440ea6f43
Signed-off-by: Tony Xie <tony.xie@rock-chips.com>
Signed-off-by: Caesar Wang <wxt@rock-chips.com>
Signed-off-by: Shengfei xu <xsf@rock-chips.com>

### CODE REVIEW

**Comment by danh-arm** [comment]
Can one of the admins verify this patch?

**Comment by danh-arm** [comment]
Can one of the admins verify this patch?

File: plat/rockchip/rk3399/drivers/pmu/pmu.c [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
```
Comment (danh-arm) [inline plat/rockchip/rk3399/drivers/pmu/pmu.c]: This license boilerplate text is not acceptable. Please use the standard boilerplate

File: plat/rockchip/rk3399/drivers/pmu/pmu.h [hunk 2]
```diff
@@ -0,0 +1,693 @@
...
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
```
Comment (danh-arm) [inline plat/rockchip/rk3399/drivers/pmu/pmu.h]: This license boilerplate text is not acceptable. Please use the standard boilerplate

File: plat/rockchip/rk3399/drivers/soc/soc.c [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
```
Comment (danh-arm) [inline plat/rockchip/rk3399/drivers/soc/soc.c]: Wrong license

File: plat/rockchip/rk3399/drivers/soc/soc.h [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
```
Comment (danh-arm) [inline plat/rockchip/rk3399/drivers/soc/soc.h]: Wrong license

**Comment by danh-arm** [comment]
jenkins: test this please

**Comment by danh-arm** [comment]
A number of outstanding issues remain from previous reviews (e.g. unimplemented system_off, use of binary blob for ddr_resume, lack of system suspend for rk3399, use of locks, ...). However, ultimately this is your responsibility so we will merge this once the license text is fixed.

**Comment by danh-arm** [comment]
jenkins: retest this please

**Comment by danh-arm** [comment]
jenkins: retest this please

File: plat/rockchip/rk3399/drivers/pmu/pmu.c (hunk 1 above)
Comment (rkchrome) [inline plat/rockchip/rk3399/drivers/pmu/pmu.c]: Done

File: plat/rockchip/rk3399/drivers/pmu/pmu.h (hunk 2 above)
Comment (rkchrome) [inline plat/rockchip/rk3399/drivers/pmu/pmu.h]: Done

File: plat/rockchip/rk3399/drivers/soc/soc.c (hunk 3 above)
Comment (rkchrome) [inline plat/rockchip/rk3399/drivers/soc/soc.c]: Done

File: plat/rockchip/rk3399/drivers/soc/soc.h (hunk 4 above)
Comment (rkchrome) [inline plat/rockchip/rk3399/drivers/soc/soc.h]: Done

**Comment by rkchrome** [comment]
The new pull on https://github.com/ARM-software/arm-trusted-firmware/pull/581

**Pull request closed without merging**